vsphere/new_hub.py

```diff
@@ -0,0 +1,210 @@
+        print("Error: {} is already in use".format(ip), file=sys.stderr)
+        sys.exit(-1)
+    elif result.returncode == 1:
+        debug("ip {} is available".format(ip))
+        return ip
+    else:
+        print("Error: {}".format(result.stderr), file=sys.stderr)
+        sys.exit(-1)
+
+def check_vips(api_vip, ingress_vip):
+    debug("Pinging api_vip: {}".format(api_vip))
+    debug("api_vip {} is available".format(ping_ip(api_vip)))
+    debug("Pinging ingress_vip: {}".format(ingress_vip))
+    debug("ingress_vip {} is available".format(ping_ip(ingress_vip)))
+
+    # FIXME: this is a hack to get the segment name, don't do this
+    segname = None
+    for name, segment in vmcdns.SEGMENTS_BY_NAME.items():
+        pattern = r'^(\d+\.\d+\.\d+)\.'
+        match = re.search(pattern, api_vip)
+        if match.group(0) in str(segment.network):
+            segname = name
+            break
+    if not segname:
+        print("Error: could not find segment name for api_vip {}".format(api_vip), file=sys.stderr)
+        sys.exit(-1)
+    return [api_vip, ingress_vip], segname
+
+def get_networks():
+    if ARGS.network:
+        return [ARGS.network]
+    elif ARGS.disconnected:
+        return [segname for segname in vmcdns.SEGMENTS_BY_NAME.keys() if "-disconnected" in segname]
+    else:
+        return [segname for segname in vmcdns.SEGMENTS_BY_NAME.keys() if "-disconnected" not in segname]
+    
+def save_install_config(ips, segname, cluster_name, creds):
+    install_config = vmcdns.get_install_config(ips, segname, cluster_name, creds)
+    # TODO check that this is a valid directory
+    filename = "install-config.yaml"
+    if ARGS.dir:
+        filename = os.path.join(ARGS.dir, "install-config.yaml")
+    with open(filename, "w") as f:
+        f.write(install_config)
+
+def create_cluster(dir):
+    try:
+        subprocess.run(["openshift-install", "create", "cluster", "--dir", dir, "--log-level=info"], check=True)
+    except subprocess.CalledProcessError as e:
+        print("ERROR: {}".format(e), file=sys.stderr)
+        print("Run the destroy command to clean up the cluster", file=sys.stderr)
+        sys.exit(-1)
+
+def destroy_cluster(dir):
+    try:
+        subprocess.run(["openshift-install", "destroy", "cluster", "--dir", dir, "--log-level=info"], check=True)
+    except subprocess.CalledProcessError as e:
+        print("ERROR: {}".format(e), file=sys.stderr)
+        sys.exit(-1)
+
+def install_hive(dir):
+    try:
+        subprocess.run(["make", "deploy"], check=True)
+    except subprocess.CalledProcessError as e:
+        print("ERROR: {}".format(e), file=sys.stderr)
+        sys.exit(-1)
+
+if __name__ == "__main__":
+    PARSER = build_parser()
+    ARGS = PARSER.parse_args()
+    CREDS = vmcdns.Creds()
+    check_args()
+
+    if ARGS.subcommand == "create-cluster":
+        avail = []
+        if ARGS.api_vip and ARGS.ingress_vip:
+           avail, segname = check_vips(ARGS.api_vip, ARGS.ingress_vip)
```

Comment:
In this path, `avail` is used for `save_install_config(avail[:2], segname, ARGs.cluster_name,CREDS)`, but if the `--api-vip` and `--ingress-vip` flags are specified, the script assumes they are already reserved and does not reserve them, though they still get pinged. The reason I refrained from having a symmetrical `cleanup-cluster` command with an option to not release, is that this would be the equivalent of simply calling `openshift-install destroy`. Do you think it's still worthwhile to add this option?

Reply:
> In this path, `avail` is used

Ah, I see it now. Whose idea was whitespace-based contexting in python? :p

> this would be the equivalent of simply calling `openshift-install destroy`.

Mm, I guess we're relying on having kept the metadata.json around anyway, huh. Okay, I'll buy it.

If this were a production utility I would be more persnickety about getting it perfect, but as long as the quirks are documented this'll work.